substate-store-impls/src/single_substate_database_overlay.rs

```diff
@@ -1,27 +1,54 @@
 use radix_engine_common::prelude::*;
 use substate_store_interface::interface::*;
+use utils::prelude::borrow::*;
 
-pub struct SingleSubstateDatabaseOverlay<'s, S> {
+pub type UnmergeableSubstateDatabaseOverlay<'a, S> = SubstateDatabaseOverlay<&'a S, S>;
+pub type MergeableSubstateDatabaseOverlay<'a, S> = SubstateDatabaseOverlay<&'a mut S, S>;
+pub type OwnedSubstateDatabaseOverlay<S> = SubstateDatabaseOverlay<S, S>;
+
+pub struct SubstateDatabaseOverlay<S, D> {
     /// The database overlay. All commits made to the database are written to the overlay. This
     /// covers new values and deletions too.
     overlay: StagingDatabaseUpdates,
 
-    /// The root database that this type overlays. There is no restriction on what type this is at
-    /// this level. This is an immutable reference to a substate store as there is no need for this
-    /// type to own the underlying database or to mutate it in any way.
-    root: &'s S,
+    /// A mutable or immutable reference to the root database that this type overlays.
+    /// It only needs to be mutable if you wish to commit to the root store.
+    /// To be useful, `S` should implement at least `Borrow<D>`.
+    root: S,
```

Comment:
Nice! My impl was missing the use of the two different generic parameters and I think that this is why I was facing issues with it. The use of `S` and `D` here is clear and quite nice!